### ISSUE TITLE
Validate connection calls after excluding null values at runtime

### ISSUE THREAD

**Issue by BerndWessels** [facebook/relay#677]
Hello
Using the following query:

      fragment on Person {
        firstName
        email
        posts(first: $first, after: $after, last: $last, before: $before) {
          edges {
            cursor
            node {
              id
              title
            }
          }
          totalCount
          pageInfo {
            hasNextPage
            hasPreviousPage
            startCursor
            endCursor
          }
        }
      }

causes the following error:

    Uncaught Error: GraphQL validation/transform error ``Connection arguments `posts(before: <cursor>, first: <count>)` are not supported. Use `(first: <count>)`, `(after: <cursor>, first: <count>)`, or `(before: <cursor>, last: <count>)`.``

So obviously some combinations of parameters are not valid - and that is fair enough.

But I was expecting that the validation only happens after it excluded variables with value `null`. Unfortunately that doesn't seem to be the case, otherwise we could use the previous query with these variables:

    export default Relay.createContainer(Home, {
      initialVariables: {
        first: 2,
        after: "YXJyYXljb25uZWN0aW9uOjQ=",
        last: null,
        before: null,
      },
      fragments: {
        viewer: () => Relay.QL`
          fragment on Person {

And it should succeed since only the `first` variable is non-null and all the other variables should be ignored / not included.

So for now we have to use terrible workarounds like this:

    export default Relay.createContainer(Home, {
      initialVariables: {
        forward: true,
        first: 2,
        after: "YXJyYXljb25uZWN0aW9uOjQ=",
        last: null,
        before: null,
      },
      fragments: {
        viewer: () => Relay.QL`
          fragment on Person {
            firstName
            email
            posts(first: $first, after: $after) @include(if: $forward) {
              edges {
                cursor
                node {
                  id
                  title
                }
              }
              totalCount
              pageInfo {
                hasNextPage
                hasPreviousPage
                startCursor
                endCursor
              }
            },
            posts(last: $last, before: $before) @skip(if: $forward) {
              edges {
                cursor
                node {
                  id
                  title
                }
              }
              totalCount
              pageInfo {
                hasNextPage
                hasPreviousPage
                startCursor
                endCursor
              }
            },
            ${UpdatePersonMutation.getFragment('person')}
          }
        `
      }
    });

So basically my feature request is that variables with `null` value will be ignored and excluded before query validation and submitting.

**Comment by josephsavona** [comment]
Yeah, this has come up before. We'd be happy to accept a PR. `GraphQLRange` should (already) have the correct logic for validating arguments after excluding nulls - we just need to confirm that, and remove the validation in RelayQLPrinter (in the plugin).

**Comment by NevilleS** [comment]
I was looking for a good way to poke around the internals of the babel-relay-plugin so I've prepared a quick PR to fix this! @josephsavona, you're correct that GraphQLRange does validate the arguments it is given at runtime, so the plugin's static validation can be loosened a bit.

I didn't end up removing the validation entirely, I just added an check to see if the arguments are variables. That means @BerndWessels's query will now be supported, but other invalid combinations using hardcoded values will still be rejected.
